packages/hooks/src/createAddHook.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * Returns a function which, when invoked, will add a hook.
+ *
+ * @param  {Object}   hooks Stored hooks, keyed by hook name.
+ *
+ * @return {Function}       Function that adds a new hook.
+ */
+function createAddHook( hooks ) {
+	/**
+	 * Adds the hook to the appropriate hooks container.
+	 *
+	 * @param {string}   hookName Name of hook to add
+	 * @param {Function} callback Function to call when the hook is run
+	 * @param {?number}  priority Priority of this hook (default=10)
+	 */
+	return function addHook( hookName, callback, priority = 10 ) {
+		if ( typeof hookName !== 'string' ) {
+			console.error( 'The hook name must be a string.' );
+			return;
+		}
+
+		if ( /^__/.test( hookName ) ) {
```

Comment:
In the interest of starting with a restrictive API that allows for expansion later, let's restrict hook names to something like `/^[a-z][a-z0-9_]/`.